File: src/shift.cpp
/*****************************************************************************
 **
 ** Copyright (C) 2019 Fabian Schweinfurth
 ** Contact: autoshift <at> derfabbi.de
 **
 ** This file is part of autoshift
 **
 ** autoshift is free software: you can redistribute it and/or modify
 ** it under the terms of the GNU Lesser General Public License as published by
 ** the Free Software Foundation, either version 3 of the License, or
 ** (at your option) any later version.
 **
 ** autoshift is distributed in the hope that it will be useful,
 ** but WITHOUT ANY WARRANTY; without even the implied warranty of
 ** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 ** GNU Lesser General Public License for more details.
 **
 ** You should have received a copy of the GNU Lesser General Public License
 ** along with autoshift.  If not, see <http://www.gnu.org/licenses/>.
 **
 *****************************************************************************/

#include <QtNetwork>
#include <QUrl>
#include <QDataStream>

#include <QDialog>

#include <misc/logger.hpp>
#include <misc/fsettings.hpp>
#include <misc/xmlreader.hpp>

#include <shift.hpp>

#include "ui_loginwindow.h"

#include <request.hpp>

#define SC ShiftClient

#define COOKIE_FILE_NAME ".cookie.sav"

const QUrl baseUrl("https://shift.gearboxsoftware.com");

const QRegularExpression rForm("(<form.*?>.*?</form>)",
                               QRegularExpression::DotMatchesEverythingOption);
const QRegularExpression rAlert("<div[^>]*class=\"[^\"]*?alert.*?\">(.*?)</div>",
                                QRegularExpression::DotMatchesEverythingOption);
const QRegularExpression rRed_status("(<div[^>]*id=\"[^\"]*?check_redemption_status.*?\">(.*?)</div>)",
                                     QRegularExpression::DotMatchesEverythingOption);

SC::ShiftClient(QObject* parent)
  :QObject(parent), logged_in(false), 
  cookieFile(QStandardPaths::writableLocation(QStandardPaths::AppLocalDataLocation)
    + "/" + COOKIE_FILE_NAME)
{
  connect(this, &ShiftClient::loggedin, this, &ShiftClient::loggedinAct);
}

SC::~ShiftClient()
{
}

void SC::loggedinAct(bool v)
{
  INFO << "login " << ((v)? "successfull" : "failed") << "!" << endl;
}

Status SC::redeem(const ShiftCode& code)
{
  StatusC formData = getRedemptionData(code);

  if (formData.code != Status::SUCCESS) {
    int status_code = formData.data.toInt();
    if (status_code == 500)
      return Status::INVALID;
    if (status_code == 429)
      return Status::SLOWDOWN;
    if (formData.message.contains("expired"))
      return Status::EXPIRED;
    if (formData.message.contains("not available"))
      return Status::UNAVAILABLE;
    // unknown
    ERROR << formData.message.trimmed() << endl;
    return Status::UNKNOWN;
  }

  QStringList lis = formData.data.value<QStringList>();
  QUrlQuery postData;
  for(size_t i = 0; i < lis.size(); i += 2) {
    postData.addQueryItem(lis[i], lis[i+1]);
  }

  StatusC status = redeemForm(postData);
  return status.code;
}

void SC::delete_cookies()
{
  auto* network_manager = static_cast<QNetworkAccessManager*>(FSETTINGS["nman"].value<void*>());
  QList<QNetworkCookie> cookies = network_manager->cookieJar()->cookiesForUrl(baseUrl);
  QByteArray data;
  for (auto& cookie: cookies) {
    network_manager->cookieJar()->deleteCookie(cookie);
  }

  QFile cookie_f(cookieFile);
  cookie_f.remove();
}

bool SC::save_cookie()
{
  auto* network_manager = static_cast<QNetworkAccessManager*>(FSETTINGS["nman"].value<void*>());
  // write cookie
  QList<QNetworkCookie> cookies = network_manager->cookieJar()->cookiesForUrl(baseUrl);
  QByteArray data;
  for (auto& cookie: cookies) {
    if (cookie.name().toStdString() == "si") {
      data = cookie.toRawForm();
      break;
    }
  }
  // no cookie found
  if (data.isEmpty()) return false;

  QFile cookie_f(cookieFile);
  if (!cookie_f.open(QIODevice::WriteOnly))
    return false;

  // write cookie to file
  int data_size = data.size();
  cookie_f.write((char*)&data_size, sizeof(int));
  cookie_f.write(data);

  // write user to file
  QString user = FSETTINGS["user"].toString();
  data_size = user.size();
  cookie_f.write((char*)&data_size, sizeof(int));
  cookie_f.write(user.toUtf8());

  cookie_f.close();

  return true;
}

bool SC::load_cookie()
{
  QFile cookie_f(cookieFile);
  if (!cookie_f.open(QIODevice::ReadOnly))
    return false;

  // read cookie
  QDataStream in(&cookie_f);
  int data_size;
  in.readRawData((char*)&data_size, sizeof(data_size));
  char* data = new char[data_size];
  in.readRawData(data, data_size);
  QByteArray cookieString(data, data_size);
  delete[] data;

  // read user
  in.readRawData((char*)&data_size, sizeof(data_size));
  data = new char[data_size];
  in.readRawData(data, data_size);
  QString user = QString::fromUtf8(data, data_size);
  FSETTINGS["user"] = user;
  delete[] data;

  cookie_f.close();

  // set cookies on current manager
  QList<QNetworkCookie> cookies = QNetworkCookie::parseCookies(cookieString);

  auto* network_manager = static_cast<QNetworkAccessManager*>(FSETTINGS["nman"].value<void*>());
  network_manager->cookieJar()->setCookiesFromUrl(cookies, baseUrl);

  // test if no success
  Request req(baseUrl.resolved(QUrl("/account")), request_t::HEAD);
  req.send();
  if (!wait(&req, &Request::finished))
    return false;

  logged_in = req.status_code == 200;
  return logged_in;
}

StatusC SC::getToken(const QUrl& url)
{
  Request req(url);
  // get request
  req.send();

  // wait for answer
  if (!wait(&req, &Request::finished)) {
    return {Status::UNKNOWN, "Timeout on Token Request"};
  }

  return getToken(req.data);
}

StatusC SC::getToken(const QString& data)
{
  StatusC ret;
  // extract token from DOM
  XMLReader xml(data);
  while (true) {
    xml.findNext("meta");
    if (xml.atEnd()) break;

    auto attrs = xml.attributes();
    auto name = attrs.value("name");

    if (name != "csrf-token") continue;

    auto token = attrs.value("content");
    if (token.isEmpty()) continue;

    return {Status::SUCCESS, token.toString()};
  }
  return {};
}

void SC::logout()
{
  Request req(baseUrl.resolved(QUrl("/logout")));
  req.send();
  wait(&req, &Request::finished);
}
void SC::login()
{
  // auto login
  if (load_cookie()) {
    logged_in = true;
    emit loggedin(logged_in);
    return;
  }

  // delete old cookies
  delete_cookies();

  // show login window
  QDialog loginDialog;
  Ui::Login loginWin;
  loginWin.setupUi(&loginDialog);

  // when clicked "Ok"
  if (loginDialog.exec()) {
    QString user = loginWin.userInput->text();
    QString pw = loginWin.pwInput->text();

    INFO << "trying to log in" << endl;
    login(user, pw);
  } else {
    emit loggedin(false);
  }
}

void SC::login(const QString& user_name, const QString& pw)
{
  FSETTINGS.setValue("user", user_name);
  // TODO redirect_to=false => wrong pw
  // get login form
  QUrl the_url = baseUrl.resolved(QUrl("/home"));

  StatusC formData = getFormData(the_url);
  if (formData.code != Status::SUCCESS) {
    emit loggedin(false);
    return;
  }

  // extract form data
  QStringList lis = formData.data.value<QStringList>();
  QUrlQuery postData;

  for(size_t i = 0; i < lis.size(); i += 2) {
    postData.addQueryItem(lis[i], lis[i+1]);
  }

  // inject our data
  postData.removeQueryItem("user[email]");
  postData.addQueryItem("user[email]", user_name);
  postData.removeQueryItem("user[password]");
  postData.addQueryItem("user[password]", pw);

  // send login request
  Request* req = new Request(baseUrl.resolved(QUrl("/sessions")),
                            postData, request_t::POST);
  // setup request
  req->req.setRawHeader("Referer", the_url.toEncoded());

  // send request and follow redirects
  req->send([&](Request* req) mutable {
    logged_in = save_cookie();
    emit loggedin(logged_in);
    req->deleteLater();
  });
}

StatusC SC::getFormData(const QUrl& url)
{
  // query site
  Request req(url);
  req.send();

  if (!wait(&req, &Request::finished)) {
    return {Status::UNKNOWN, "Could not query " + url.toString()};
  }

  return getFormData(req.data, Platform::NONE);
}

StatusC SC::getFormData(const QString& data, Platform platform)
{
  QStringList formData;

  // TODO: filter for the services (psn, xbox, epic, steam)
  //    <input value="epic" type="hidden" name="archway_code_redemption[service]" id="archway_code_redemption_service">
  // only parse body as the parser doesn't like <meta .. > tags
  QString form;
  QRegularExpressionMatchIterator i = rForm.globalMatch(data);
  QString searchPlatform = QString::fromStdString(sPlatform(platform)).toLower();

  while (i.hasNext()) {
	  QRegularExpressionMatch match = i.next();
	  std::string tmp = match.captured(1).toStdString();
	  if (platform == Platform::NONE) {
		  form = match.captured(1);
		  break;
	  }
	  XMLReader xml(match.captured(1));
    bool found = xml.findNextWithAttr("input", "value", searchPlatform);

	  if (found) {
		  form = match.captured(1);
		  break;
	  }
  }

  if (form.isEmpty())
    return {Status::UNKNOWN, data};

  // find form on url
  XMLReader xml(form);

  // find input fields
  while (true) {
    bool tmp = xml.readNextStartElement();
    if (xml.atEnd()) break;
    // are we at the </form> tag?
    if (!tmp && xml.name() == "form")
      break;
    // continue on end-tags or non-input tags
    if (!tmp || xml.name() != "input")
      continue;

    auto attrs = xml.attributes();
    // collect name and value
    formData << attrs.value("name").toString() << attrs.value("value").toString();
  }

  return {Status::SUCCESS, form, formData};
}

StatusC SC::getRedemptionData(const ShiftCode& code)
{
  StatusC ret;

  StatusC s = getToken(baseUrl.resolved(QUrl("/code_redemptions/new")));

  if (s.code != Status::SUCCESS) {
    ERROR << "Token Request Error: " << s.message << endl;
    return ret;
  }

  QString& token(s.message);

  // get form
  Request req(baseUrl.resolved(
                QUrl(QString("/entitlement_offer_codes?code=%1").arg(code.code()))));
  req.req.setRawHeader("x-csrf-token", token.toUtf8());
  req.req.setRawHeader("x-requested-with", "XMLHttpRequest");

  // send and follow redirections
  req.followRedirects(true);

  req.send();
  if (!wait(&req, &Request::finished))
    return ret;

  ret.data = req.reply->attribute(QNetworkRequest::HttpStatusCodeAttribute);

  // now extract form-data
  StatusC formData = getFormData(req.data, code.platform());

  ret.code = formData.code;
  ret.message = formData.message;

  if (formData.code == Status::SUCCESS) {
    ret.data = formData.data;
  }

  return ret;
}

StatusC SC::getAlert(const QString& content)
{
  auto match = rAlert.match(content);
  if (!match.hasMatch())
    return {};

  QString alert = match.captured(1).trimmed();

  QString alert_l = alert.toLower();
  Status status = Status::NONE;
  if (alert_l.contains("to continue to redeem")) {
    status = Status::TRYLATER;
  } else if (alert_l.contains("expired")) {
    status = Status::EXPIRED;
  } else if (alert_l.contains("success")) {
    status = Status::SUCCESS;
  } else if (alert_l.contains("failed")) {
    status = Status::REDEEMED;
  }

  return {status, alert};
}

StatusC SC::getStatus(const QString& content)
{
  auto match = rRed_status.match(content);
  if (!match.hasMatch())
    return {};

  QString div = match.captured(1);
  QString text = match.captured(2).trimmed();

  // find form on url
  XMLReader xml(div);
  xml.readNextStartElement();

  QStringList urls;
  urls << xml.attributes().value("data-url").toString();
  urls << xml.attributes().value("data-fallback-url").toString();
  return {Status::REDIRECT, text, urls};
}

StatusC SC::checkRedemptionStatus(const Request& req)
{
  // redirection
  if (req.status_code == 302)
    return {Status::REDIRECT, req.reply->rawHeader("location")};

  // check if there is a alert box..
  StatusC alert = getAlert(req.data);
  if (alert.code != Status::NONE) {
    INFO << alert.message << endl;
    return alert;
  }

  StatusC status = getStatus(req.data);
  if (status.code == Status::REDIRECT) {
    // "follow" redirects
    QStringList urls = status.data.value<QStringList>();
    QString new_location = urls[0];
    QString token = getToken(req.data).message;
    QString msg;

    // only try 5 times
    for (uint8_t cnt=0; ; ++cnt) {
      // query for status
      if (cnt > 5) {
        return {Status::REDIRECT, urls[1]};
      }

      if (msg != status.message) {
        // don't spam..
        msg = status.message;
        INFO << status.message << endl;
      }

      // query url
      Request new_req(baseUrl.resolved(new_location));
      new_req.req.setRawHeader("x-csrf-token", token.toUtf8());
      new_req.req.setRawHeader("x-requested-with", "XMLHttpRequest");
      new_req.send();
      wait(&new_req, &Request::finished);

      // extract json and generate usable information
      QJsonDocument json = QJsonDocument::fromJson(new_req.data);
      if (json["text"] != QJsonValue::Undefined) {
        QString json_text = json["text"].toString();
        if (json_text.contains("success"))
          return {Status::SUCCESS, json_text};
        if (json_text.contains("failed"))
          return {Status::REDEEMED, json_text};
      }

      // wait for 500ms (wait for a signal that won't fire...)
      wait(&req, &Request::finished, 500);
    }

  }
  return {};
}

StatusC SC::redeemForm(const QUrlQuery& data)
{
  // cache old rewards
  // old_rewards = queryRewards();
  QUrl the_url = baseUrl.resolved(QUrl("/code_redemptions"));
  // bool redemption = false;

  StatusC status;
  request_t req_type = request_t::POST;
  do {
  Request req(the_url, data, req_type);
  req.req.setRawHeader("Referer", baseUrl.resolved(QUrl("/code_redemptions/new")).toEncoded());

  req.send();
  if (!wait(&req, &Request::finished))
    return {Status::UNKNOWN, "timed out"};

  status = checkRedemptionStatus(req);
  if (status.code == Status::REDIRECT) {
    the_url = baseUrl.resolved(QUrl(status.message));
    if (status.message.contains("home?redirect_to"))
      return {Status::UNKNOWN, "Something weird happened..."};
  }

  req_type = request_t::GET;
  } while (status.code == Status::REDIRECT);

  return status;
}

#undef SC

File: autoshift/shift.py
#############################################################################
#
# Copyright (C) 2018 Fabian Schweinfurth
# Contact: autoshift <at> derfabbi.de
#
# This file is part of autoshift
#
# autoshift is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# autoshift is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with autoshift.  If not, see <http://www.gnu.org/licenses/>.
#
#############################################################################

import pickle
import time
from enum import Enum
from typing import Any, Literal

import httpx
import typer
from bs4 import BeautifulSoup as BSoup
from bs4 import Tag

from autoshift.common import _L, settings
from autoshift.models import Key

base_url = "https://shift.gearboxsoftware.com"


def json_headers(token: str) -> dict[str, str]:
    return {"x-csrf-token": token, "x-requested-with": "XMLHttpRequest"}


# filthy enum hack with auto convert
class Status(Enum):
    NONE = "Something unexpected happened.."
    REDIRECT = "<target location>"
    TRYLATER = (
        "To continue to redeem SHiFT codes, please launch a SHiFT-enabled title first!"
    )
    EXPIRED = "This code expired by now.. ({key.reward})"
    REDEEMED = "Already redeemed {key.reward}"
    SUCCESS = "Redeemed {key.reward}"
    INVALID = "The code `{key.code}` is invalid"
    SLOWDOWN = "Too many requests"
    UNKNOWN = "An unknown Error occured: {msg}"

    def __init__(self, s: str) -> None:
        self.msg = s

    @classmethod
    def _missing_(cls, value: object):
        obj = object.__new__(cls)
        obj._value_ = value
        obj.__init__(str(value))
        cls._value2member_map_[value] = obj  # type: ignore # these bindings are wrong..
        return obj

    def __eq__(self, other: Any) -> bool:
        if not isinstance(other, self.__class__):
            return False
        return self._name_ == other._name_

    def __call__(self, new_msg: str) -> "Status":
        if "{msg}" in new_msg:
            new_msg = new_msg.format(msg=new_msg)
        obj = self.__class__(new_msg)
        for k, v in vars(self).items():
            if not hasattr(obj, k):
                setattr(obj, k, v)
        return obj


class ShiftClient:
    logged_in: bool = False

    def __init__(self):
        # try to load cookies. Query for login data if not present
        self.cookies = self.__load_cookie()
        self.client = httpx.Client(follow_redirects=True, cookies=self.cookies)

    def login(self, user: str | None = None, pw: str | None = None):
        if self.cookies:
            self.logged_in = self.check_login()
        if self.logged_in:
            return True
        typer.echo("Login to your SHiFT account...")
        if not user:
            user = typer.prompt("Username")
        if not pw:
            pw = typer.prompt("Password", hide_input=True)

        if user and pw:
            self.__login(user, pw)
        if self.__save_cookie():
            _L.info("Login Successful")
            self.logged_in = True
        else:
            _L.error("Couldn't login. Are your credentials correct?")
            exit(0)

    def check_login(self) -> bool:
        response = self.client.get(f"{base_url}/rewards")
        return response.status_code == 200 and "Sign Out" in response.text

    def __save_cookie(self) -> bool:
        """Make ./data folder if not present"""
        if not settings.COOKIE_FILE.parent.exists():
            settings.COOKIE_FILE.parent.mkdir(parents=True, exist_ok=True)

        self.client.cookies.jar.clear_expired_cookies()

        """Save cookie for auto login"""
        with settings.COOKIE_FILE.open("wb") as f:
            if "si" in self.client.cookies:
                pickle.dump(self.client.cookies.jar._cookies, f)  # pyright: ignore[reportAttributeAccessIssue]
                return True
        return False

    def __load_cookie(self) -> httpx.Cookies | None:
        """Check if there is a saved cookie and load it."""
        if not settings.COOKIE_FILE.exists() or settings.COOKIE_FILE.stat().st_size == 0:
            return None
        try:
            cookies = httpx.Cookies()
            with settings.COOKIE_FILE.open("rb") as f:
                jar_cookies = pickle.load(f)
                for domain, pc in jar_cookies.items():
                    for path, c in pc.items():
                        for k, v in c.items():
                            cookies.set(k, v.value, domain=domain, path=path)
            return cookies
        except Exception:
            _L.error("Could not load cookies. Re-login required")
            return None

    def redeem(self, key: Key) -> Status:
        from time import sleep

        retry = True
        status = Status.NONE
        game_longname = key.game.long_name

        while retry:
            found, status_code, form_data = self.__get_redemption_form(
                key.code, game_longname, key.platform
            )
            # the expired message comes from even wanting to redeem
            if not found:
                if status_code >= 500:
                    # entered key was invalid
                    status = Status.INVALID
                elif status_code == 429:
                    status = Status.SLOWDOWN
                elif "expired" in form_data:
                    status = Status.EXPIRED
                elif "not available" in form_data:
                    status = Status.INVALID
                elif "does not exist" in form_data:
                    status = Status.INVALID
                elif "already been redeemed" in form_data:
                    status = Status.REDEEMED
                else:
                    # unknown
                    # _L.error(form_data)
                    status = Status.UNKNOWN(str(form_data))
            else:
                # the key is valid and all.
                if isinstance(form_data, dict):
                    status = self.__redeem_form(form_data)
                else:
                    status = Status.UNKNOWN(str(form_data))

            if status == Status.SLOWDOWN:
                sleep(60)
            else:
                retry = False

            key.redeemed = status in (Status.SUCCESS, Status.REDEEMED)
            key.expired = status in (Status.EXPIRED, Status.INVALID)
            key.save()

            if key.expired:
                # set all keys with the same code as expired
                (Key.update(expired=True).where(Key.code == key.code).execute())

        return status

    def __get_token(self, url_or_reply: str | httpx.Response) -> tuple[int, str | None]:
        """Get CSRF-Token from given URL"""
        if isinstance(url_or_reply, str):
            r = self.client.get(url_or_reply)
        else:
            r = url_or_reply

        soup = BSoup(r.text, "html.parser")
        meta = soup.find("meta", attrs=dict(name="csrf-token"))
        if not isinstance(meta, Tag):
            return r.status_code, None
        return r.status_code, str(meta.get("content", ""))

    def __login(self, user: str, pw: str) -> httpx.Response | None:
        """Login with user/pw"""
        the_url = f"{base_url}/home"
        _, token = self.__get_token(the_url)
        if not token:
            return None
        login_data = {
            "authenticity_token": token,
            "user[email]": user,
            "user[password]": pw,
        }
        headers = {"Referer": the_url}
        r = self.client.post(f"{base_url}/sessions", data=login_data, headers=headers)
        _L.debug(f"{r.request.method} {r.url} {r.status_code}")
        return r

    def __get_redemption_form(
        self, code: str, game: str | None, platform: str
    ) -> tuple[Literal[False], int, str] | tuple[Literal[True], int, dict[str, str]]:
        """Get Form data for code redemption"""

        the_url = f"{base_url}/rewards"
        status_code, token = self.__get_token(the_url)
        if not token:
            _L.debug("no token")
            return False, status_code, "Could not retrieve Token"

        r = self.client.get(
            f"{base_url}/entitlement_offer_codes?code={code}", headers=json_headers(token)
        )

        if r.status_code != 200:
            return False, r.status_code, str(r.status_code)

        soup = BSoup(r.text, "html.parser")
        if not soup.find("form", class_="new_archway_code_redemption"):
            return False, r.status_code, r.text.strip()

        titles = soup.find_all("h2")
        title = titles[0]
        # some codes work for multiple games and yield multiple buttons for the same platform..
        if len(titles) > 1:
            for _t in titles:
                if isinstance(_t, Tag) and _t.text == game:
                    title = _t
                    break

        if not isinstance(title, Tag):
            return (False, r.status_code, "Could not find title tag")

        forms = title.find_all_next("form", id="new_archway_code_redemption")
        for form in forms:
            if not isinstance(form, Tag):
                continue
            service = form.find(id="archway_code_redemption_service")
            if isinstance(service, Tag) and platform in str(service.get("value", "")):
                inputs = form.find_all("input")
                form_data = {}
                for inp in inputs:
                    if isinstance(inp, Tag):
                        name = inp.get("name")
                        value = inp.get("value")
                        if name:
                            form_data[str(name)] = str(value or "")
                return True, r.status_code, form_data

        return (False, r.status_code, "This code is not available for your platform")

    def __get_status(self, alert: str) -> Status:
        status = Status.NONE
        lower = alert.lower()
        if "success" in lower:
            status = Status.SUCCESS
        elif "failed" in lower or "been redeemed" in lower:
            status = Status.REDEEMED

        return status

    def __get_redemption_status(self, r: httpx.Response) -> tuple[str, str, str]:
        # return None
        soup = BSoup(r.text, "lxml")
        div = soup.find("div", id="check_redemption_status")
        if not div:
            div = soup.find("div", class_="alert notice")
        if isinstance(div, Tag):
            return (
                div.text.strip(),
                str(div.get("data-url", "")),
                str(div.get("data-fallback-url", "")),
            )
        return ("", "", "")

    def __check_redemption_status(self, r: httpx.Response) -> Status:
        """Check redemption"""
        import json
        from time import sleep

        if r.status_code == 302:
            return Status.REDIRECT(r.headers["location"])

        get_status, url, fallback = self.__get_redemption_status(r)
        if get_status:
            if not url:
                return self.__get_status(get_status)

            _, token = self.__get_token(r)
            cnt = 0
            # follow all redirects
            while True:
                if cnt > 5:
                    return Status.REDIRECT(fallback)
                _L.info(get_status)
                _L.debug(f"get {base_url}/{url}")
                raw_json = self.client.get(
                    f"{base_url}/{url}",
                    follow_redirects=False,
                    headers=json_headers(token or ""),
                )
                _L.debug(f"Raw json text: {raw_json.text}")
                data = json.loads(raw_json.text)

                if "text" in data:
                    return self.__get_status(data["text"])
                # wait 500
                sleep(0.5)
                cnt += 1

        return Status.NONE

    def _query_rewards(self) -> list[str]:
        """Query reward list"""
        # self.old_rewards
        the_url = f"{base_url}/rewards"
        r = self.client.get(the_url)
        soup = BSoup(r.text, "html.parser")

        # cache all unlocked rewards
        return [el.text for el in soup.find_all("div", class_="reward_unlocked")]

    def __redeem_form(self, data: dict[str, str]) -> Status:
        """Redeem a code with given form data"""

        the_url = f"{base_url}/code_redemptions"
        headers = {"Referer": f"{base_url}/rewards"}
        response = self.client.post(
            the_url, data=data, headers=headers, follow_redirects=False
        )
        _L.debug(f"{response.request.method} {response.url} {response.status_code}")
        status = self.__check_redemption_status(response)
        # did we visit /code_redemptions/...... route?
        redemption = False
        # keep following redirects
        while status == Status.REDIRECT:
            if "code_redemptions/" in status.value:
                redemption = True
                while True:
                    response2 = self.client.get(
                        status.value,
                        headers={
                            "referer": status.value,
                            "x-requested-with": "XMLHttpRequest",
                            "accept": "application/json",
                        },
                    )
                    json_data = response2.json()
                    if json_data.get("in_progress", False):
                        time.sleep(0.5)
                        continue
                    try:
                        status = self.__get_status(json_data["text"])
                    except KeyError as e:
                        _L.error("Unexpected JSON response. Please report this issue!")
                        _L.error(f"JSON data: {json_data}")
                        raise e
                    break

            else:
                _L.debug(f"redirect to '{status.value}'")
                response2 = self.client.get(status.value)
                status = self.__check_redemption_status(response2)

        # workaround for new SHiFT website.
        # it doesn't tell you to launch a "SHiFT-enabled title" anymore
        if status == Status.NONE:
            if redemption:
                status = Status.REDEEMED
            else:
                status = Status.TRYLATER
        return status
